### ISSUE TITLE
Can't easily run junit tests in buildSrc

### ISSUE THREAD

**Issue by cmccabe** [gradle/gradle#8224]
When test files are placed into the buildSrc/ directory, it is difficult to run them from the command line.

Expected behavior:  I would expect `gradle buildSrc:test --tests MyTest` to run MyTest.java in the buildSrc directory.

Actual behavior: Gradle complains that there is no such subproject as "buildSrc."  While this is technically correct, it is frustrating since gradle displays :buildSrc:test as a phase in the command-line output, suggesting that tasks in that phase can be run like any others.  I cannot find any documentation about how tests in buildSrc are supposed to be run.

How has this issue affected you: we have chosen to not put any tests into buildSrc since we can't figure out how to run them from the command line.  Instead we put them in a subproject and use custom sourceSets to include the buildSrc code in the test context.

Steps to Reproduce: create any simple project with a buildSrc directory and a test named MyTest.java.  The test file can't be run via `gradle buildSrc:test --tests=MyTest`

Your Environment: Arch Linux

**Comment by cmccabe** [comment]
Hi @marcphilipp,

Thanks for the response.  Is this documented in the manual, so that others can discover how to run these tests?

**Comment by cmccabe** [comment]
This solution doesn't work.  When I try to compile inside the buildSrc directory I get:
```
FAILURE: Build failed with an exception.

* Where:
Build file '/home/cmccabe/t/kafka/buildSrc/build.gradle' line: 21

* What went wrong:
A problem occurred evaluating root project 'buildSrc'.
> Could not find method compile() for arguments [com.fasterxml.jackson.core:jackson-databind:2.9.6] on object of type org.gradle.api.internal.artifacts.dsl.dependencies.DefaultDependencyHandler.
```

If I add `apply plugin:'java'`, then I get:
```
> Task :compileJava FAILED                                                                                                                                                                                                                                                                
/home/cmccabe/src/kafka1/buildSrc/src/main/java/org/apache/kafka/task/ProcessMessagesTask.java:21: error: package org.gradle.api does not exist                                                                                                                                           
import org.gradle.api.DefaultTask;                                                                                                                                                                                                                                                        
...
```

You can try it out yourself.  Check out https://github.com/apache/kafka and try building inside `buildSrc`.

**Comment by marcphilipp** [comment]
Have you tried adding an empty `settings.gradle[.kts]` to `buildSrc`?

I agree that this needs to be documented better.

**Comment by eskatos** [comment]
It seems to me that there's also the `buildSrc/build.gradle` that is missing some plugins and dependencies: https://github.com/apache/kafka/blob/trunk/buildSrc/build.gradle

It works from the outer build because then the `groovy` plugin is automatically applied and the `gradleApi()` is automatically added by Gradle, this is a bit confusing.

@cmccabe please try to add a `buildSrc/settings.gradle` file and the following to your `buildSrc/build.gradle`:

```groovy
plugins {
    id("groovy")
}

dependencies {
    compile(gradleApi())
}
```

**Comment by Sanvijaya** [comment]
As suggested in below link, you can use 'gradle -p buildSrc test'
It works for me.

https://discuss.gradle.org/t/running-buildsrc-tasks-from-root-project/22069

**Comment by jjohannes** [comment]
With Gradle 7.0 you can use an included build for build logic and run a task in that build directly.

Instead of `buildSrc` use another folder - e.g. `build-logic`. Then include that in your settings.gradle file.

```
pluginManagement {
  includeBuild("build-logic")
}
```

Then you can run e.g.:

```
gradle :build-logic:test --tests MyTest 
```

**Comment by autonomousapps** [comment]
@jjohannes when you include the build like that, does that let you create a settings plugin in that included build, and use it in the main build?

**Comment by jjohannes** [comment]
Yes. See: https://docs.gradle.org/7.0/release-notes.html#plugin-development-improvements

**Comment by cmccabe** [comment]
Thanks for looking at this again.

In Kafka we just decided to create a normal subproject called "generator" which generates the source files.  I'm not honestly sure if using "buildSrc" has any advantages over doing this -- if so, it needs to be documented better in the future before I'd recommend it to other people, I think.

**Comment by cmccabe** [comment]
> This issue has been automatically marked as stale because it has not had recent activity. Given the limited bandwidth of the team, it will be automatically closed if no further activity occurs. 

here is a comment to prevent the issue from being automatically closed

**Comment by adammurdoch** [comment]
This is fixed in Gradle 8.0 via https://github.com/gradle/gradle/pull/22540. You can now run `buildSrc` tasks directly from the command line, eg `./gradlew buildSrc:test`
